Development practices/folder_structure.md

```diff
@@ -0,0 +1,286 @@
+│   ├── settler.rb
+│   ├── status_setter.rb
+│   ├── updater.rb
+│   ├── users_setter.rb
+│   └── voider.rb
+├── shared
+│   └── filter_suggested.rb
+├── stats_maker.rb
+├── subscriptions
+│   └── actions.rb
+├── tln
+│   └── number_checker.rb
+├── translators
+│   └── google.rb
+├── units.rb
+├── update_user_to_developer.rb
+├── user_loyalty_points.rb
+├── users
+│   ├── actions.rb
+│   ├── creator.rb
+│   ├── destroyer.rb
+│   ├── exporter.rb
+│   ├── o_auth_creator.rb
+│   ├── retrieve_all_data
+│   │   ├── base.rb
+│   │   ├── favorites.rb
+│   │   ├── friends_list.rb
+│   │   ├── notifications.rb
+│   │   ├── personal_information.rb
+│   │   └── reviews.rb
+│   ├── retrieve_all_data.rb
+│   └── settings.rb
+├── vat_checker.rb
+├── version_converter.rb
+└── x_server
+    ├── add_detour_distance.rb
+    ├── add_driving_distance.rb
+    ├── base.rb
+    ├── build_polyline.rb
+    ├── corridor_search.rb
+    ├── get_detour_distance.rb
+    ├── get_driving_distance.rb
+    ├── get_geo_keywords.rb
+    ├── get_location_attributes.rb
+    ├── get_route.rb
+    └── leave_reachable.rb
+```
+
+</details>
+
+At some point this folder became a dumpall folder without rhyme or reason.
+There is also an issue of boundries: there are none. Each class is used from anywhere in the application and that can
+cause headaches.
+
+# Desired structure
+
+Here is a folder structure we should strive for
+
+```
+app
+├── controllers  # Rails 
+├── frontend     # webpacker
+├── domains      # custom
+├── models       # Rails
+├── views        # Rails
+└── web          # custom
+```
+
+Lets first talk about Rails specific folders (`controllers`, `views` and `models`).
+
+`views` folder needs to exist because it is hardcoded in the Rails itself (https://github.com/rails/rails/blob/main/railties/lib/rails/engine.rb#L610)
+
+`controllers` is there because of the way rails finds controller names (it appends `"Controller"` to the resource name) (https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/http/request.rb#L88)
+
+`models` is a folder where we put all our ActiveRecord models. It is technically not necessary but we will talk about it in [Further Improvements](#further-improvements) chapter.
+
+
+# `domains` vs `web` folder
+`web` is for classes used in controller actions: forms, policies, serializers, queries and similar.
+`domains` is for business logic.
+
+## `web` folder
+`web` folder structure should follow your routes. Having these routes:
+
+```ruby
+namespace: :dasboard do
+  resources :projects
+end
+namespace :api do
+  namespace :v1 do
+    resources :projects
+  end
+end
+```
+
+it would have this folder structure:
+
+```
+web
+├── dashboard
+│   ├── projects
+│   │   ├── presenter.rb
+│   │   ├── decorator.rb
+│   │   ├── form.rb
+│   │   └── policy.rb
+├── api
+│   ├── v1
+│   │   ├── v1
+│   │   │   ├── projects
+│   │   │   │   ├── policy.rb
+│   │   │   │   ├── query.rb
+│   │   │   │   └── serialzier.rb
```

Comment:
typo

Reply:
Can you also add example of how we moved mailers folder into web folder?
https://github.com/infinum/rails-jsonapi-example-app/tree/master/app/web/mailers

I really like that approach because I could create a decorator that is used for a specific resource only in mailers (bookings resource)

Reply:
I actually added mailers quite sneakily here: https://github.com/infinum/rails-handbook/pull/130/files#diff-d5c0a8a84263c854da60332d0552ceb26b600f950c547cff123e6b4dd8b22b91R269
in the domains folder

because sending a specific mail is usually done in one place, its not something thats reused throughout the application. For example `Mailers::Bookings::Mailer` is used only in one of the controller. My point is that it does not need to have a `Mailers` namespace, they can just live where where they are used.

Although i guess this needs to be tested. There might be an issue of finding the right template files. Drat.